modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_7_3/UpgradeDDMFormInstanceReport.java

```diff
@@ -0,0 +1,416 @@
+							"formInstanceRecordId");
+
+						DDMFormValues ddmFormValues = _getDDMFormValues(
+							data, ddmForm);
+
+						dataJSONObject = _processDDMFormValues(
+							dataJSONObject, ddmFormValues,
+							formInstanceRecordId);
+
+						dataJSONObject.put(
+							"totalItems",
+							dataJSONObject.getInt("totalItems") + 1);
+					}
+				}
+
+				long groupId = rs1.getLong("groupId");
+
+				long companyId = rs1.getLong("companyid");
+
+				Timestamp createDate = rs1.getTimestamp("createDate");
+
+				try (PreparedStatement ps3 =
+						AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+							connection, sb2.toString())) {
+
+					ps3.setLong(1, increment());
+					ps3.setLong(2, groupId);
+					ps3.setLong(3, companyId);
+					ps3.setTimestamp(4, createDate);
+					ps3.setTimestamp(5, createDate);
+					ps3.setLong(6, formInstanceId);
+					ps3.setString(7, dataJSONObject.toString());
+
+					ps3.execute();
+				}
+			}
+		}
+	}
+
+	private DDMForm _getDDMForm(long structureId) throws Exception {
+		try (PreparedStatement ps = connection.prepareStatement(
+				"select parentStructureId, definition, storageType from " +
+					"DDMStructure where structureId = ?")) {
+
+			ps.setLong(1, structureId);
+
+			try (ResultSet rs = ps.executeQuery()) {
+				if (rs.next()) {
+					String definition = rs.getString("definition");
+
+					DDMFormDeserializerDeserializeRequest.Builder builder =
+						DDMFormDeserializerDeserializeRequest.Builder.
+							newBuilder(definition);
+
+					DDMFormDeserializerDeserializeResponse
+						ddmFormDeserializerDeserializeResponse =
+							_ddmFormDeserializer.deserialize(builder.build());
+
+					return ddmFormDeserializerDeserializeResponse.getDDMForm();
+				}
+			}
+
+			throw new UpgradeException(
+				"Unable to find dynamic data mapping structure with ID " +
+					structureId);
+		}
+	}
+
+	private DDMFormValues _getDDMFormValues(String data, DDMForm ddmForm)
+		throws Exception {
+
+		DDMFormValues ddmFormValues = new DDMFormValues(ddmForm);
+
+		List<DDMFormFieldValue> ddmFormFieldValues = new ArrayList<>();
+
+		JSONObject dataJSONObject = _jsonFactory.createJSONObject(data);
+
+		JSONArray fieldValuesJSONArray = dataJSONObject.getJSONArray(
+			"fieldValues");
+
+		Iterator<JSONObject> iterator = fieldValuesJSONArray.iterator();
+
+		while (iterator.hasNext()) {
+			JSONObject jsonObject = iterator.next();
+
+			String name = jsonObject.getString("name");
+
+			DDMFormFieldValue ddmFormFieldValue = new DDMFormFieldValue();
+
+			ddmFormFieldValue.setDDMFormValues(ddmFormValues);
+			ddmFormFieldValue.setInstanceId(jsonObject.getString("instanceId"));
+			ddmFormFieldValue.setName(name);
+
+			DDMFormField ddmFormField = ddmFormFieldValue.getDDMFormField();
+
+			Value value = null;
+
+			if (ddmFormField.isLocalizable()) {
+				value = new LocalizedValue();
+
+				JSONObject valueJSONObject = jsonObject.getJSONObject("value");
+
+				for (String key : valueJSONObject.keySet()) {
```

Comment:
@marcosapmf, this is consistent, but if we rename `key` to `languageId` can be more meaningful. Do you think we should change? What's your opinion?

Reply:
@natocesarrego, I agree with you! I am going to make this change in the next pull :)